OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -766,16 +766,52 @@ void TracerThread::ProcessExitEvent(const perf_event_header& header,
 
 void TracerThread::ProcessMmapEvent(const perf_event_header& header,
                                     PerfEventRingBuffer* ring_buffer) {
-  pid_t pid = ReadMmapRecordPid(ring_buffer);
+  // Mmap records have the following layout:
+  // struct {
+  //   struct perf_event_header header;
+  //   u32    pid, tid;
+  //   u64    addr;
+  //   u64    len;
+  //   u64    pgoff;
+  //   char   filename[];
+  //   struct sample_id sample_id; /* if sample_id_all */
+  // };
+  // Because of filename, the layout is not fixed.
+
+  perf_event_sample_id_tid_time_streamid_cpu sample_id;
+  ReadPerfSampleIdAll(ring_buffer, header, &sample_id);
+
+  perf_event_mmap_up_to_filename mmap_event;
+  ring_buffer->ReadValueAtOffset(&mmap_event, 0);
+
+  //  ring_buffer->ConsumeRecord()
+  // read filename
+  size_t filename_offset = sizeof(perf_event_mmap_up_to_filename);
+  // strictly > because filename is null-terminated string
+  CHECK(header.size > (filename_offset + sizeof(perf_event_sample_id_tid_time_streamid_cpu)));
+  size_t filename_size =
+      header.size - filename_offset - sizeof(perf_event_sample_id_tid_time_streamid_cpu);
+  std::vector<char> filename_vector(filename_size);
+  ring_buffer->ReadRawAtOffset(&filename_vector[0], filename_offset, filename_size);
+  std::string filename(&filename_vector[0], filename_size);
+
   ring_buffer->SkipRecord(header);
```

Comment:
Move this code to a `ConsumeMmapPerfEvent` in `PerfEventReaders.h`, which is precisely where we have functions to read `PerfEvent`s that are not trivial to read.

It's fine to create the whole `MapsPerfEvent` before the `if`s that follow, it's a rare event anyway.